### ISSUE TITLE
Decoding URI GeneralNames doesn't handle IPv6 addresses

### ISSUE THREAD

**Issue by vpelletier** [pyca/cryptography#3863]
To reproduce:
- generate a certificate with:
```python
      builder = builder.add_extension(
        x509.CRLDistributionPoints([
          x509.DistributionPoint(
            full_name=[
              x509.UniformResourceIdentifier(u'http://[::1]:80/some/path'),
            ],
            relative_name=None,
            crl_issuer=None,
            reasons=None,
          ),
        ]),
        critical=False,
      )
```
- serialise certificate: works
- load certificate: works
- iterate over extensions: raises when `cryptography` encounters above extension.

Here is the tail of the traceback I get:

```
  File ".../cryptography-2.1.dev1-py2.7-linux-x86_64.egg/cryptography/utils.py", line 159, in inner
    result = func(instance)
  File ".../cryptography-2.1.dev1-py2.7-linux-x86_64.egg/cryptography/hazmat/backends/openssl/x509.py", line 134, in extensions
    self._backend, self._x509
  File ".../cryptography-2.1.dev1-py2.7-linux-x86_64.egg/cryptography/hazmat/backends/openssl/decode_asn1.py", line 233, in parse
    value = handler(backend, ext_data)
  File ".../cryptography-2.1.dev1-py2.7-linux-x86_64.egg/cryptography/hazmat/backends/openssl/decode_asn1.py", line 544, in _decode_crl_distribution_points
    backend, cdp.distpoint.name.fullname
  File ".../cryptography-2.1.dev1-py2.7-linux-x86_64.egg/cryptography/hazmat/backends/openssl/decode_asn1.py", line 83, in _decode_general_names
    names.append(_decode_general_name(backend, gn))
  File ".../cryptography-2.1.dev1-py2.7-linux-x86_64.egg/cryptography/hazmat/backends/openssl/decode_asn1.py", line 96, in _decode_general_name
    hostname = idna.decode(parsed.hostname)
  File "build/bdist.linux-x86_64/egg/idna/core.py", line 384, in decode
    result.append(ulabel(label))
  File "build/bdist.linux-x86_64/egg/idna/core.py", line 299, in ulabel
    check_label(label)
  File "build/bdist.linux-x86_64/egg/idna/core.py", line 253, in check_label
    raise InvalidCodepoint('Codepoint {0} at position {1} of {2} not allowed'.format(_unot(cp_value), pos+1, repr(label)))
InvalidCodepoint: Codepoint U+003A at position 1 of u'::1' not allowed
```

On the other hand, replacing the IPv6 value with `127.0.0.1` or `localhost` works.

**Comment by alex** [comment]
Updating the title to reflect the underlying problem, this isn't CRL specific.

**Comment by vpelletier** [comment]
Researching some more, I believe idna is doing its job:

- [rfc5891 section-4.2.2](https://tools.ietf.org/html/rfc5891#section-4.2.2) says:

  ```
  The candidate Unicode string MUST NOT contain characters that appear
  in the "DISALLOWED" and "UNASSIGNED" lists specified in the Tables
  document [RFC5892].
  ```

- [rfc5892 appendix B.1](https://tools.ietf.org/html/rfc5892#appendix-B.1) does contain:

  ```
  003A..0060  ; DISALLOWED  # COLON..GRAVE ACCENT
  ```

So ":", which is U+003A, must be rejected.

Although "." (U+002E) is also forbidden, it is obviously used to split components of domain names, so 127.0.0.1 is actually parsed as a domain name (1.2.3 and 1.2.3.4.5 pass too).

**Comment by reaperhulk** [comment]
I haven't confirmed but it might only error in master if you access `.value` on the URI (and `.bytes_value` will hopefully give what you expect).

**Comment by vpelletier** [comment]
According to [rfc3986 section-3.2.2](https://tools.ietf.org/html/rfc3986#section-3.2.2), what `urlparse` exposes as `netloc` may contain:

- `[` &lt;stuff&gt; `]`: literal IP (IPv6 but potentially more)
- dotted-quad IPv4
- domain name

The RFC acknowldeges the ambiguity between dotted-quad and domain name, telling to use first matches with dotted-quad before domain name.

So now I believe this test should be done before calling idna, to not give it literal IPs (not even IPv4).

Then comes the question of how much matching should happen before dispatching. What about something like:

```python
if parsed.hostname[0] == "[" and "]" in parsed.hostname:
    # TODO: support optional "vX." prefix ?
    ipaddress.ip_address(
        parsed.hostname[1:].rsplit("]", 1)[0].decode("ascii")
    )
    hostname = parsed.hostname
else:
    try:
        ipaddress.ip_address(parsed.hostname.decode("ascii"))
    except ValueError:
        hostname = idna.decode(parsed.hostname)
    else:
        hostname = parsed.hostname
```

**Comment by reaperhulk** [comment]
This should be resolved with #3905 and then accessing the value of the parsed data via `bytes_value`
